File: Metropolis Algorithm/metropolis.jl
using Distributions
using Random

""" # Configuration for MCMC
    MCMC_Config([chain::Int, init::Real, jump, posterior, space::AbstractVector, burnin::Int, rng]) -> MCMC_Config
* `chain::Int` - the length of Markov Chain to be generated (default 10000)
* `init::Real` - initial value for the chain of θ (default 0.5)
* `jump` - the distribution that proposed jump (Δθ) follows (default Distributions.Normal(0,0.2))
* `posterior` 
    - (`::Symbol`) the Symbol of self-defined density function
    - (`::var(#)`) a anonymous function of the density function (default θ -> Distributions.pdf(Beta(15,7),θ))
* `space::AbstractVector` - the accepted span of the parameter θ [min,max] (default [0,1])
* `burnin::Int` - the length of the chain to be burned from the beginning (default 0)
* `rng` - random seed (default 42), set `rng = nothing` to use no random seed
"""
Base.@kwdef struct MCMC_Config
    chain::Int = 10000
    init::Real = 0.5
    jump = Distributions.Normal(0,0.2)
    posterior = θ -> Distributions.pdf(Beta(15,7),θ) # posterior = :p
    space::AbstractVector = [0,1]
    burnin::Int = 0
    rng = 42
end
""" # Metropolis Algorithm
    metropolis(dfunc, chain::Int, theta_init::Real; [jump = Normal(0,0.2), space_min::Real = -Inf, space_max::Real = Inf, burnin::Int = 0, rng = nothing]) -> AbstractVector

Compute a Markov Chain using Metropolis Algorithm

# Parameters
## arguments
* `dfunc`
    - (`::Symbol`) the Symbol of self-defined density function
    - (`::var(#)`) a anonymous function of the density function
* `chain::Int` - the length of Markov Chain to be generated
* `theta_init::Real` - initial value for the chain of θ (default 0.5)

## keyword arguments
* `jump` - the distribution that proposed jump (Δθ) follows (default Distributions.Normal(0,0.2))
* `space_min::Real` - the accepted minimum of the parameter θ (default -Inf)
* `space_max::Real` - the accepted maximum of the parameter θ (default  Inf)
* `burnin::Int` - the length of the chain to be burned from the beginning (default 0)
* `rng` - random seed (default nothing), set `rng = 0` to use seed `MersenneTwister(0)`

# Examples
```julia-repo
julia> p(θ) = Distributions.pdf(Beta(15,7),θ) # define a density function

julia> metropolis(:p, 5000, 0.5; jump = Normal(0,0.2), space_min = 0, space_max = 1, burnin = 1000, rng = 42)
4000-element Vector{Float64}:
 0.7254870615709366
 0.5374079418869023
 0.5374079418869023
 0.5161750364065404
 0.7377560837277216
 0.7377560837277216
 0.8127557482035582
 ⋮
 0.7166820320569407
 0.7166820320569407
 0.7166820320569407
 0.7166820320569407
 0.7166820320569407
 0.7586353644413187

julia> @time metropolis(θ -> Distributions.pdf(Beta(15,7),θ), 10000, 0.5; jump = Normal(0,0.5), space_min = 0, space_max = 1, burnin = 1000, rng = nothing)
0.061715 seconds (103.38 k allocations: 3.814 MiB, 88.90% compilation time)
9000-element Vector{Float64}:
 0.6057362032299476
 0.6057362032299476
 0.6057362032299476
 0.6057362032299476
 0.6057362032299476
 0.6057362032299476
 0.6057362032299476
 ⋮
 0.4516105573548591
 0.4516105573548591
 0.4516105573548591
 0.5014099540247418
 0.6720304713998495
 0.6720304713998495

```
"""
function metropolis(dfunc, chain::Int, theta_init::Real; jump = Normal(0,0.2), space_min::Real = -Inf, space_max::Real = Inf, burnin::Int = 0, rng = nothing)
    posterior = 0 # initialize posterior to the outer scope
    try
        if typeof(dfunc) == Symbol
            posterior = getfield(Main,dfunc)
        else
            posterior = dfunc
        end
    catch
        print("dfunc must be a self-defined function name or an anonymous")
    end

    θ_cur = theta_init
    θ_freq = [theta_init]

    Random.seed!(MersenneTwister(rng))

    while true
        Δθ = rand(jump)
        θ_pro = θ_cur + Δθ

        if θ_pro < space_min || θ_pro > space_max
            pmoving = 0
        elseif posterior(θ_cur) == 0
            pmoving = 1
        else
            pmoving = min(1, posterior(θ_pro)/posterior(θ_cur))
        end

        if rand() <= pmoving
            push!(θ_freq, θ_pro)
            θ_cur = θ_pro
        else
            push!(θ_freq, θ_cur)
        end

        if length(θ_freq) >= chain
            break
        end
    end

    return θ_freq[burnin + 1 : end]
end

""" 
    metropolis(Config::MCMC_Config) -> AbstractVector

Compute a Markov Chain using Metropolis Algorithm

use the struct constructor `::MCMC_Config` to configure the parameter for Metropolis Algorithm

check struct `MCMC_Config` for more details

# Examples
```julia-repo
julia> Config = MCMC_Config()

julia> metropolis(Config)
10000-element Vector{Float64}:
 0.5
 0.6828984833360351
 0.6828984833360351
 0.6655360165668127
 0.6655360165668127
 0.6655360165668127
 0.6591963728158979
 ⋮
 0.7246337711515528
 0.5807104209871898
 0.5275540714270037
 0.5569823211525173
 0.5272108581003658
 0.5272108581003658

julia> p(θ) = Distributions.pdf(Gamma(1,2),θ) # define a density function

julia> Config = MCMC_Config(chain = 5000, jump = Distributions.TDist(5), space = [0,Inf], posterior = :p, burnin = 500, rng = nothing)

julia> metropolis(Config)
4500-element Vector{Float64}:
 0.34362723284778773
 0.34362723284778773
 1.1023030285120496
 1.1023030285120496
 1.1023030285120496
 1.1023030285120496
 1.565816384278301
 ⋮
 2.05294789570401
 0.80709730394856
 0.80709730394856
 0.9931268934687827
 0.9931268934687827
 0.9931268934687827

```
"""
function metropolis(Config::MCMC_Config)
    dfunc = Config.posterior
    chain = Config.chain
    theta_init = Config.init
    jump = Config.jump
    space_min = Config.space[1]
    space_max = Config.space[2]
    burnin = Config.burnin
    rng = Config.rng
    metropolis(dfunc, chain, theta_init; jump = jump, space_min = space_min, space_max = space_max, burnin = burnin, rng = rng)
end

File: Metropolis Algorithm/metropolis.py
import numpy as np
from scipy.special import beta
import scipy.stats

class MCMC:
    """Markov Chain Monte Carlo method class used by MCMC Algorithms functions (e.g. metropolis)

    Used to initialize instances for MCMC Algorithm functions

    Return an MCMC object

    Parameters
    ----------
    dfunc : function
        self-defined density function which take only one parameter (i.e. r.v. X)
    chain : int, optional
        the length of Markov Chain to be generated (default 5000)
    theta_init : float, optional
        initial value for the chain of θ (default 0.5)
    jumpdist : scipy.stats.rv_continuous or scipy.stats.rv_discrete, optional
        the distribution that proposed jump (Δθ) follows (default scipy.stats.norm(0,0.2))
    space : list, optional
        the accepted span of the parameter θ (default [-np.inf,np.inf])
    burnin : int, optional
        the length of the chain to be burned from the beginning (default 0)
    seed : int or None, optional
        random seed (default None)

    Usage:
    -----
    >>> import metropolis
    >>> density = lambda x: scipy.stats.gamma(2,loc=4,scale=5).pdf(x)
    >>> d = metropolis.MCMC(density, chain = 10000, jumpdist=scipy.stats.norm(loc=0,scale=2), space = [0,np.inf])
    >>> d
    <metropolis.MCMC object ...>
    """
    def __init__(self, 
                 dfunc, 
                 chain = 5000, 
                 theta_init = 0.5, 
                 jumpdist = scipy.stats.norm(0,0.2), 
                 space = [-np.inf,np.inf], 
                 burnin = 0, 
                 seed = None) -> None:
        self.dfunc = dfunc
        self.chain = chain
        self.theta_init = theta_init
        self.jump = jumpdist
        self.space = space
        self.burnin = burnin
        self.seed = seed

    def metropolis(self, *arg, **kwarg):
        """
        function applying Metropolis Algorithm to MCMC objects

        Parameters
        ----------
        dfunc : function, *arg, optional (update)
            self-defined density function which take only one parameter (i.e. r.v. X)
        chain : int, optional (update)
            the length of Markov Chain to be generated
        theta_init : float, optional (update)
            initial value for the chain of θ (default 0.5)
        jumpdist : scipy.stats.rv_continuous or scipy.stats.rv_discrete, optional (update)
            the distribution that proposed jump (Δθ) follows (default scipy.stats.norm(0,0.2))
        space : list, optional (update)
            the accepted span of the parameter θ (default [-np.inf,np.inf])
        burnin : int, optional (update)
            the length of the chain to be burned from the beginning (default 0)
        seed : int or None, optional (update)
            random seed (default None)

        Examples
        --------
        >>> import metropolis
        >>> from scipy.special import beta
        >>> density = lambda x: x**14 * (1-x)**6 / beta(15,7) # Beta(15,7) distribution density function
        >>> d = metropolis.MCMC(density, space = [0,1], burnin = 5, seed = 72)
        >>> result = d.metropolis(chain = 50, theta_init = 0.1)
        >>> result
        [0.8176960093922774, 0.6965658096789994, 0.7918980615882665, 0.7742394795862185, 0.7742394795862185, 0.6215414421599866, 0.6215414421599866, 0.6468248283946754, 0.7523307146724492, 0.7523307146724492, 0.7680822171469903, 0.6717376155310788, 0.6717376155310788, 0.8189878864825765, 0.7533257832372013, 0.7648206889254298, 0.7648206889254298, 0.7648206889254298, 0.7648206889254298, 0.7648206889254298, 0.7967947965010628, 0.707139903476412, 0.707139903476412, 0.707139903476412, 0.707139903476412, 0.7949590848197712, 0.48018105229278457, 0.48018105229278457, 0.6163978253871556, 0.6163978253871556, 0.6163978253871556, 0.6241841537823003, 0.6241841537823003, 0.6241841537823003, 0.6241841537823003, 0.6922332333961135, 0.8204027203103916, 0.7521267229207833, 0.7521267229207833, 0.808566620237602, 0.808566620237602, 0.6460288022318678, 0.5452360032045938, 0.5452360032045938, 0.5934357613729606]
        """
        # update attributes
        for args in arg:
            self.dfunc = args
        
        for kw, value in kwarg.items():
            if kw == 'chain':
                self.chain = value
            elif kw == 'theta_init':
                self.theta_init = value
            elif kw == 'jumpdist':
                self.jump = value
            elif kw == 'space':
                self.space = value
            elif kw == 'burnin':
                self.burnin = value
            elif kw == 'seed':
                self.seed = value
            else:
                raise Exception(f'keyword argument "{kw}" not supported',)

        # check if dfunc callable
        if not callable(self.dfunc):
            raise Exception("dfunc must be a function. recreate the object with a valid density function")
        
        # Metropolis Algorithm
        theta_cur = self.theta_init
        theta_freq = [self.theta_init]
        
        rng = np.random.default_rng(self.seed)

        while True:
            Delta_theta = self.jump.rvs(random_state=rng)
            theta_pro = theta_cur + Delta_theta

            if theta_pro < self.space[0] or theta_pro > self.space[1]:
                pmoving = 0
            elif self.dfunc(theta_cur) == 0:
                pmoving = 1
            else:
                pmoving = min(1,self.dfunc(theta_pro)/self.dfunc(theta_cur))
            
            # np.random.rand()
            if scipy.stats.uniform().rvs(random_state=rng) <= pmoving:
                theta_freq.append(theta_pro)
                theta_cur = theta_pro
            else:
                theta_freq.append(theta_cur)

            if len(theta_freq) >= self.chain:
                break

        return theta_freq[self.burnin:]

def main():
    # density = lambda x: x**14 * (1-x)**6 / beta(15,7)
    density = lambda x: scipy.stats.gamma(2,loc=4,scale=5).pdf(x)
    d = MCMC(density, chain = 10, jumpdist=scipy.stats.norm(loc=0,scale=2), space = [0,np.inf])
    print(d)
    result = d.metropolis(chain=100, seed = 42)
    print(result)
    
if __name__ == "__main__":
    main()